integration/app/faucet_test.go

```diff
@@ -0,0 +1,136 @@
+package app
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+	envtest "github.com/tendermint/starport/integration"
+	"github.com/tendermint/starport/starport/chainconfig"
+	"github.com/tendermint/starport/starport/pkg/cosmosfaucet"
+	"github.com/tendermint/starport/starport/pkg/xurl"
+	"golang.org/x/sync/errgroup"
+	"io"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+)
+
+const (
+	addr = "cosmos1zqr2gd7hwkyw55knad0l6ml6ngutd70878evqj"
+)
+
+var (
+	defaultCoins = []string{"10token", "1stake"}
+	maxCoins = []string{"102token", "100000000stake"}
+)
+
+type QueryAllBalancesResponse struct {
+	Balances sdk.Coins `json:"balances"`
+}
+
+func TestRequestCoinsFromFaucet(t *testing.T) {
+	var (
+		env               = envtest.New(t)
+		apath             = env.Scaffold("faucet")
+		servers           = env.RandomizeServerPorts(apath, "")
+		faucetURL            = env.ConfigureFaucet(apath, "", defaultCoins, maxCoins)
+		ctx, cancel       = context.WithTimeout(env.Ctx(), envtest.ServeTimeout)
+	)
+	// serve the app
+	go func() {
+		env.Must(env.Serve("should serve app", apath, "", "", envtest.ExecCtx(ctx)))
+	}()
+
+	// wait servers to be online
+	defer cancel()
+	err := env.IsAppServed(ctx, servers)
+	require.NoError(t, err)
+
+	err = env.IsFaucetServed(ctx, faucetURL)
+	require.NoError(t, err)
+
+	// error "account doesn't have any balances" occurs if a sleep is not included
+	time.Sleep(time.Second*1)
```

Comment:
Instead of sleep(s), can we backoff.Retry?

Reply:
I use a backoff here for the faucet but it still fails with the error mentioned in the comment without adding a delay

They are also sleeps between faucet requests and balance reads because the faucet request sends the tx but doesn't wait for it to be confirmed. I may try to remove them after finishing the PR but it may potentially but out of scope of the PR if `chainrunner` has no simple way to wait for the tx to be mined

Reply:
the disadvantage with the sleep line here is, we cannot be sure that tx will be mined after sleep has done. that's why we need a retry mechanism.

Reply:
f51fb6db92a143657fd79fb6f013eef0c62a18ad

Reply:
Though I couldn't remove the `Sleep` highlighted above. There is still a little delay after faucet is online for the requests to succeed